File: fw-portal-web/src/main/java/com/faraway/fwportal/dto/ConhecimentoDto.java
package com.faraway.fwportal.dto;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ConcurrentSkipListSet;

import com.faraway.fwportal.model.Conhecimento;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.fasterxml.jackson.annotation.JsonRootName;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;

@JsonRootName("response")
@JsonPropertyOrder({ "chave", "numero", "serie", "emissao", "total", "taxas", "imposto", "emitente", "remetente",
		"destinatario", "origem", "destino", "carga", "notas" })
@JsonInclude(Include.NON_NULL)
public class ConhecimentoDto {

	@JacksonXmlProperty(isAttribute = true, localName = "chaveCte")
	@JsonProperty("chaveCte")
	private String chave;
	private String numero;
	private String serie;
	private BigDecimal total;

	@JsonFormat(pattern = "dd/MM/yyyy")
	private LocalDate emissao;

	private TransportadoraDto emitente;

	private EmpresaDto remetente;

	private EmpresaDto destinatario;

	private CidadeDto origem;

	private CidadeDto destino;

	@JsonProperty("camponentesPrest")
	private Set<TaxaConhecimentoDto> taxas;

	private ImpostoDto imposto;

	@JsonProperty("infoCarga")
	private CargaDto carga;

	private Set<NotaDto> notas;

	public ConhecimentoDto(Conhecimento conhecimento) {
		super();
		this.chave = conhecimento.getChave();
		this.numero = conhecimento.getNumero();
		this.serie = conhecimento.getSerie();
		this.total = conhecimento.getTotal();
		this.emissao = conhecimento.getEmissao();
		this.emitente = new TransportadoraDto(conhecimento.getEmitente());
		this.remetente = new EmpresaDto(conhecimento.getRemetente());
		this.destinatario = new EmpresaDto(conhecimento.getDestinatario());
		this.origem = new CidadeDto(conhecimento.getCidadeInicio());
		this.destino = new CidadeDto(conhecimento.getCidadeDestino());
		this.taxas = conhecimento.getTaxas().parallelStream().map(TaxaConhecimentoDto::new)
				.collect(ConcurrentSkipListSet::new, Set::add, Set::addAll);
		this.imposto = new ImpostoDto(conhecimento.getImposto());
		this.carga = new CargaDto(conhecimento.getCarga());
		this.notas = conhecimento.getNotas().parallelStream().map(NotaDto::new).collect(ConcurrentSkipListSet::new,
				Set::add, Set::addAll);
	}

	public String getChave() {
		return chave;
	}

	public String getNumero() {
		return numero;
	}

	public String getSerie() {
		return serie;
	}

	public BigDecimal getTotal() {
		return total;
	}

	public LocalDate getEmissao() {
		return emissao;
	}

	public TransportadoraDto getEmitente() {
		return emitente;
	}

	public EmpresaDto getRemetente() {
		return remetente;
	}

	public EmpresaDto getDestinatario() {
		return destinatario;
	}

	public CidadeDto getOrigem() {
		return origem;
	}

	public CidadeDto getDestino() {
		return destino;
	}

	public Set<TaxaConhecimentoDto> getTaxas() {
		return new HashSet<TaxaConhecimentoDto>(taxas);
	}

	public ImpostoDto getImposto() {
		return imposto;
	}

	public CargaDto getCarga() {
		return carga;
	}

	public Set<NotaDto> getNotas() {
		return notas;
	}

}

File: fw-portal-web/src/main/resources/static/resources/js/dto/ConhecimentoDto.js
class ConhecimentoDto {

    _numero;
    _serie;
    _chave;
    _emissao;
    _total;
    _taxas=[];
    _imposto;
    _emitente;
    _remetente;
    _destinatario;
    _origem;
    _destino;
    _carga;
    _notas=[];

    constructor(numero,serie,chave,emissao,total,taxas,imposto,emitente,remetente,
        destinatario,origem,destino,carga,notas){
        this._numero = numero;
        this._serie = serie;
        this._chave = chave;
        this._emissao = emissao;
        this._total =total;
        this._taxas = taxas.map(function(object){
            var nomeTaxa = object.nome;
            var valor = object.valor;
            return new TaxaDto(nomeTaxa,valor);
        });
      
        this.createImpostoObject(imposto);
        this._emitente = this.createEntidadeObject(emitente);
        this._remetente =  this.createEntidadeObject(remetente);
        this._destinatario =  this.createEntidadeObject(destinatario);
        this._origem = this.creatCidadeObject(origem);
        this._destino = this.creatCidadeObject(destino);
        this._carga = new CargaDto(carga.valor,carga.produtoPredominante,carga.medidas);
        this._notas = notas.map(function(object){
              return new NotaDto(object.numero,object.serie,object.chave);      
        })
        
     }

   
    createImpostoObject(imposto){
        this._imposto = new ImpostoDto(imposto.baseCalculo,imposto.porcentagemIcms,imposto.valorIcms);
    }

    createEntidadeObject(entidade){
            return new EntidadeDto(entidade.CNPJ,entidade.nome);
       
    }

    creatCidadeObject(cidade){
        return new CidadeDto(cidade.nome,cidade.uf,cidade.codigo);
        }
    
    getNumero() {
        return this._numero;
    }
    
    getSerie() {
        return this._serie;
    }
    
    getChave() {
        return this._chave;
    }
    getEmissao(){
        return this._emissao;
    }
    getTotal(){
        return this._total;
    }

    getTaxas(){
        return this._taxas;
    }

    getImposto(){
        return this._imposto;
    }

    getEmitente(){
        return this._emitente;
    }

    getRemetente(){
        return this._remetente;
    }

    getDestinatario(){
        return this._destinatario;
    }

    getOrigem(){
        return this._origem;
    }

    getDestino(){
        return this._destino;
    }

    getCarga(){
        return this._carga;
    }

    getNotas(){
        return this._notas;
    }
}